### PR TITLE
Internalize sequence combinator to controller.

### LINKED ISSUE #992: Provide sequence ID in addition to activation IDs
To track the behavior of a sequence (e.g. when debugging going through logs) comprised of multiple activations the same way one would use a transaction or session ID on other products customers suggested the introduction of a sequence ID being associated with each action invoked as part of a particular sequence.

Thoughts?

### LINKED ISSUE #995: Discussion: Sequence abort convention
@perryibm @rabbah and I discussed sequences. I'm summarizing the conclusions of the discussion.

User-visible aspects:
- A sequence will have an activation id
- The logs of a sequence are constructed by the OW runtime as follows: each line contains an activation id and the result (success, failure) for the activation id
- Entitlement check is performed at the moment of invocation for all actions (i.e., to invoke a sequence, the user has to have the rights to invoke each action in the sequence). There won't be any partial execution of sequences (e.g., first 2 actions in the sequence are executed as the user has rights, while the remaining ones are not if the user doesn't have the right to invoke the third action in the sequence).

Implementation internals:
- The current implementation with `pipe.js` in the invoker will stay in place but guarded by a flag for experimental/comparison reasons
- The logic of executing a sequence are moved to the controller. The controller starts an actor each time it needs to execute a sequence. The actor executes in blocking fashion each action by posting directly to Kafka.
- The logs for the sequence are constructed as explained above and stored in the activation record for the sequence.
- The annotation for the activation will contain a field kind that indicates this activation was a sequence

More comments from @rabbah 
- entitlement check for the sequence component should be done both at create/update and invoke time.
- note: the actor doesn't perform an entitlement check before invoking - it's the route handler that does; the actor simply executes the invokes, waits for responses, and then writes the activation record.
- the result of the sequence should be the result of the last action in the sequence.

### LINKED ISSUE #1253: Write method for package resolution and intermediate entities in package bindings
Write a method that given a package produces the fully resolved package name wrt to package bindings. Write a method that given a package produces all the entities for which entitlement needs to be checked (intermediary packages due to bindings).

Refactor existing code to use these methods.

### PR DESCRIPTION
**Pull request by rabbah**
Squashed all of #1473 and fixing some defects.
Fixes #995.
Fixes #992.
Fixes #1253.

### CODE REVIEW

File: core/controller/src/main/scala/whisk/core/controller/SequenceActions.scala [hunk 1]
```diff
@@ -0,0 +1,399 @@
...
+        action: WhiskAction,
+        activationId: ActivationId,
+        wskActivationEithers: Vector[Either[ActivationResponse, WhiskActivation]],
+        topmost: Boolean,
+        cause: Option[ActivationId],
+        start: Instant,
+        end: Instant): WhiskActivation = {
+
+        // extract all successful activations from the vector of activation eithers
+        val (right, left) = wskActivationEithers.span(_.isRight)
```
Comment (markusthoemmes) [inline core/controller/src/main/scala/whisk/core/controller/SequenceActions.scala]: Are you sure `span` is the right choice vs. `partition`?

From the scala doc:

> Splits this list into a prefix/suffix pair according to a predicate.

As I read that, `right` in your case is not guaranteed to only contain rights, which will result in an exception in the next line.

File: core/controller/src/main/scala/whisk/core/controller/SequenceActions.scala [hunk 2]
```diff
@@ -0,0 +1,400 @@
...
+                    // if the future fails with a timeout, the failure is dealt with at the caller level
+                    futureSeqResultTimeout map { wskActivationTuples =>
+                        val wskActivationEithers = wskActivationTuples.map(_._1)
+                        // the execution of the sequence was successful, return the result
+                        val end = Instant.now(Clock.systemUTC())
+                        val seqActivation = Some(makeSequenceActivation(user, action, seqActivationId, wskActivationEithers, topmost, cause, start, end))
+                        val atomicActionCnt = wskActivationTuples.last._2
+                        (seqActivationId, seqActivation, atomicActionCnt)
+                    } andThen {
+                        case Success((_, id, _)) => seqActivationPromise.success(id)
```
Comment (ioana-blue) [inline core/controller/src/main/scala/whisk/core/controller/SequenceActions.scala]: `id` is really an activation, so `activation` would be less confusing.

File: core/controller/src/main/scala/whisk/core/controller/SequenceActions.scala [hunk 3]
```diff
@@ -0,0 +1,400 @@
...
+                    val end = Instant.now(Clock.systemUTC())
+                    // the response of the sequence is the response of the very last activation
+                    makeSequenceActivation(user, action, seqActivationId, wskActivationEithers, topmost, cause, start, end)
+                }
+                storeSequenceActivation(activation)
+            case Failure(t: Throwable) =>
+                // consider this whisk error
+                // TODO shall we attempt storing the activation if it exists or even inspect the futures?
+                // this should be a pretty serious whisk errror if it gets here
+                error(this, s"Sequence activation 'seqActivationId' failed: ${t.getMessage}")
```
Comment (ioana-blue) [inline core/controller/src/main/scala/whisk/core/controller/SequenceActions.scala]: missing '$'

**Comment by rabbah** [comment]
pg2/610.

**Comment by psuter** [comment]
Merged as ca15c68d348a2a02cf9da54475e96b43d48a3dac

**Comment by rabbah** [comment]
@ioana-blue up to you to close #992, #995 and #1253.

**Pull request closed without merging**